lib/test-utils/test/Test/QuickCheck/ExtraSpec.hs

```diff
@@ -590,6 +621,88 @@ prop_selectMapEntries_union m0 (Positive (Small i)) =
             "number of selected entries = 0" $
         Map.fromList kvs `Map.union` m1 === m0
 
+--------------------------------------------------------------------------------
+-- Repeatedly shrinking while a condition holds
+--------------------------------------------------------------------------------
+
+prop_shrinkWhile_coverage :: Arbitrary a => Fun a Bool -> a -> Property
+prop_shrinkWhile_coverage (applyFun -> condition) a
+    = checkCoverage
+    $ cover 10
+        (isJust shrinkWhileResult)
+        "isJust shrinkWhileResult"
+    $ cover 10
+        (isNothing shrinkWhileResult)
+        "isNothing shrinkWhileResult"
+    $ property True
+  where
+    shrinkWhileResult = shrinkWhile condition shrink a
+
+prop_shrinkWhile_isNothing :: Arbitrary a => Fun a Bool -> a -> Property
+prop_shrinkWhile_isNothing (applyFun -> condition) a =
+    isNothing (shrinkWhile condition shrink a)
+    === (not (condition a) || null (L.find condition (shrink a)))
+
+prop_shrinkWhile_satisfy :: Arbitrary a => Fun a Bool -> a -> Property
+prop_shrinkWhile_satisfy (applyFun -> condition) a =
+    all condition (shrinkWhile condition shrink a)
+    === True
+
+prop_shrinkWhileSteps_coverage :: Arbitrary a => Fun a Bool -> a -> Property
+prop_shrinkWhileSteps_coverage (applyFun -> condition) a
+    = checkCoverage
+    $ cover 10
+        (null shrinkWhileStepsResult)
+        "null shrinkWhileStepsResult"
+    $ cover 10
+        (length shrinkWhileStepsResult == 1)
+        "length shrinkWhileStepsResult == 1"
+    $ cover 10
+        (length shrinkWhileStepsResult >= 2)
+        "length shrinkWhileStepsResult >= 2"
+    $ property True
+  where
+    shrinkWhileStepsResult = shrinkWhileSteps condition shrink a
+
+prop_shrinkWhileSteps_satisfy :: Arbitrary a => Fun a Bool -> a -> Property
+prop_shrinkWhileSteps_satisfy (applyFun -> condition) a =
+    all condition (shrinkWhileSteps condition shrink a)
+    === True
+
+unit_shrinkWhileSteps_Int :: Spec
+unit_shrinkWhileSteps_Int = unitTests
+    "unit_shrinkWhileSteps_Int"
+    (\(a, condition) -> shrinkWhileSteps condition shrink a)
+    tests
+  where
+    tests :: [UnitTestData (Int, (Int -> Bool)) [Int]]
+    tests =
+        [ UnitTestData
+            { params = (1024, (>= 0))
+            , result = [0]
```

Comment:
Shouldn't this list be something like:

```
            , result = [512, 256, 128, 64, 32, 16, 8, 4, 2, 1, 0]
```

Don't all those values satisfy the condition (>= 0)?

Reply:
> Shouldn't this list be something like:
> ```
>             , result = [512, 256, 128, 64, 32, 16, 8, 4, 2, 1, 0]
> ```

Not quite!

At each step, the `shrinkWhileSteps` function yields the smallest matching shrink that it can find.

You can see this by comparing the following two examples:

## Example 1

In this sequence, the smallest value produced by `shrink` just happens to match the test condition:
```hs
>>> shrink 1024
[0,512,768,896,960,992,1008,1016,1020,1022,1023]

>>> shrinkWhileSteps (>= 0) shrink 1024
[0]
```

Step-by-step analysis:
```hs
>>> find (>= 0) (shrink 1024)
Just 0
>>> find (>= 0) (shrink 0)
Nothing
```
Since `0` cannot be shrunk further, we terminate there.

## Example 2

In this sequence, the smallest value produced by `shrink` does not match the test condition, so we're forced to pick the next smallest value that does. This reveals the exponential back-off strategy of the `shrink @Int` function:

```hs
>>> shrink 1024
[0,512,768,896,960,992,1008,1016,1020,1022,1023]

>>> shrinkWhileSteps (>= 100) shrink 1024
[512,256,128,112,105,102,101,100]
```

Step-by-step analysis:
```hs
>>> find (>= 100) (shrink 1024)
Just 512
>>> find (>= 100) (shrink 512)
Just 256
>>> find (>= 100) (shrink 256)
Just 128
>>> find (>= 100) (shrink 128)
Just 112
>>> find (>= 100) (shrink 112)
Just 105
>>> find (>= 100) (shrink 105)
Just 102
>>> find (>= 100) (shrink 102)
Just 101
>>> find (>= 100) (shrink 101)
Just 100
>>> find (>= 100) (shrink 100)
Nothing
```

Reply:
One way to think about `shrinkWhileSteps` is that it generates a "slice" of the shrink space for a given value and shrinker, where all the values in that "slice" satisfy the condition.

Reply:
@sevanspowell I've added a clarification to the comments for `shrinkWhile` and `shrinkWhileSteps` here: 302bf78d074f69dca45d9a25ffd8daf6e2f8599e